File: scripts/lib/parse-citations.ts
/**
 * Extracts legal citations from bill text.
 *
 * Intentionally conservative — prefers false negatives over false positives
 * to keep referenced_laws arrays clean.
 */

// "42 U.S.C. 7401", "42 U.S.C. § 7401", "42 USC 7401"
const USC_PATTERN = /(\d+)\s*U\.?S\.?C\.?\s*§?\s*(\d+[a-z]?(?:[-–]\d+[a-z]?)?)/gi

// "Clean Air Act", "Higher Education Act of 1965"
// Requires ≥2 capitalized words before "Act" to filter generic references
const ACT_PATTERN = /(?:the\s+)?((?:[A-Z][a-zA-Z]*\s+){1,6}Act(?:\s+of\s+\d{4})?)/g

// "Public Law 117-169", "P.L. 117-169"
const PL_PATTERN = /(?:Public\s+Law|P\.?\s*L\.?)\s*(\d{1,3})[-–](\d{1,4})/gi

// Acts that are generic references, not specific law names
const ACT_BLOCKLIST = new Set([
  'This Act', 'That Act', 'The Act', 'An Act', 'Such Act',
  'Any Act', 'Each Act', 'Every Act',
])

export interface BillCitations {
  uscSections: string[]  // ["42 USC 7401", "20 USC 1001"]
  actNames: string[]     // ["Clean Air Act", "Higher Education Act of 1965"]
  publicLaws: string[]   // ["PL 117-169"]
}

export function extractCitations(text: string): BillCitations {
  const uscSections = new Set<string>()
  const actNames = new Set<string>()
  const publicLaws = new Set<string>()

  let match: RegExpExecArray | null

  USC_PATTERN.lastIndex = 0
  while ((match = USC_PATTERN.exec(text)) !== null) {
    uscSections.add(`${match[1]} USC ${match[2]}`)
  }

  ACT_PATTERN.lastIndex = 0
  while ((match = ACT_PATTERN.exec(text)) !== null) {
    const name = match[1].trim()
    if (!ACT_BLOCKLIST.has(name) && name.split(/\s+/).length >= 2) {
      actNames.add(name)
    }
  }

  PL_PATTERN.lastIndex = 0
  while ((match = PL_PATTERN.exec(text)) !== null) {
    publicLaws.add(`PL ${match[1]}-${match[2]}`)
  }

  return {
    uscSections: [...uscSections].sort(),
    actNames: [...actNames].sort(),
    publicLaws: [...publicLaws].sort(),
  }
}

File: pipeline/lib/parse_citations.py
"""
Extracts legal citations from bill text.
Port of scripts/lib/parse-citations.ts

Intentionally conservative — prefers false negatives over false positives
to keep referenced_laws arrays clean.
"""
import re
from dataclasses import dataclass, field

# "42 U.S.C. 7401", "42 U.S.C. § 7401", "42 USC 7401"
USC_PATTERN = re.compile(r"(\d+)\s*U\.?S\.?C\.?\s*§?\s*(\d+[a-z]?(?:[-\u2013]\d+[a-z]?)?)", re.IGNORECASE)

# "Clean Air Act", "Higher Education Act of 1965"
ACT_PATTERN = re.compile(r"(?:the\s+)?((?:[A-Z][a-zA-Z]*\s+){1,6}Act(?:\s+of\s+\d{4})?)")

# "Public Law 117-169", "P.L. 117-169"
PL_PATTERN = re.compile(r"(?:Public\s+Law|P\.?\s*L\.?)\s*(\d{1,3})[-\u2013](\d{1,4})", re.IGNORECASE)

ACT_BLOCKLIST = {
    "This Act", "That Act", "The Act", "An Act", "Such Act",
    "Any Act", "Each Act", "Every Act",
}


@dataclass
class BillCitations:
    usc_sections: list[str] = field(default_factory=list)  # ["42 USC 7401"]
    act_names: list[str] = field(default_factory=list)      # ["Clean Air Act"]
    public_laws: list[str] = field(default_factory=list)    # ["PL 117-169"]


def extract_citations(text: str) -> BillCitations:
    usc_sections: set[str] = set()
    act_names: set[str] = set()
    public_laws: set[str] = set()

    for m in USC_PATTERN.finditer(text):
        usc_sections.add(f"{m.group(1)} USC {m.group(2)}")

    for m in ACT_PATTERN.finditer(text):
        name = m.group(1).strip()
        if name not in ACT_BLOCKLIST and len(name.split()) >= 2:
            act_names.add(name)

    for m in PL_PATTERN.finditer(text):
        public_laws.add(f"PL {m.group(1)}-{m.group(2)}")

    return BillCitations(
        usc_sections=sorted(usc_sections),
        act_names=sorted(act_names),
        public_laws=sorted(public_laws),
    )
